### ISSUE TITLE
Set up new Solr server with more memory

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#2590]
We have 64GB right now, but it's pretty much all in use. We can switch to a machine with twice the RAM for not much more money. Here's the comparison of what we have now vs what I think we should switch to:

| | Instance Type | Hourly Cost | CPU Count | Memory
-- | -- | -- | -- | --
**We have** | m6a.4xlarge | $0.6912 | 16 | 64 GiB
**We should use** | r6g.4xlarge | $0.8064 | 16 | 128 GiB

I think we'll be able to downsize this when we switch to Elastic Search, but until then we need to at least handle the traffic we have. 

This shouldn't be too hard because the Solr index and root filesystems are on an Elastic File System. I should be able to spin up the new machine, launch it, then move the traffic over.

**Comment by mlissner** [comment]
Well, it looks like we can't do this until July without losing some money. We committed to using m6a instance for one year, so if we bail now, we've effectively already paid for it. 

I *think* we can wait until July to do this if I keep an eye on the memory utilization. I set up an alert for that if it goes above 90%. Hopefully that'll be enough. I'm also meeting with an AWS cost optimization person next week; maybe they'll have ideas. 

Whatever happens, a few notes from my day's learning:

 - It's possible we can't attach the EBS volumes to the new system simultaneously, but I think we can. There are docs on this, but I haven't read them and was just going to give it a try.
 - When we do this, we'll need to update the alerts in CloudWatch to trigger for the new machine.
 - I think once the new machine is ready, we should stop the old, start the new, and deploy a change to SOLR_URL. We should avoid having two Solr instances using the same disk at the same time. I assume that'd cause chaos and don't want to find out.

**Comment by mlissner** [comment]
Well, shoot. I thought the memory usage would be more gradual, but it's not. Another thing we can do is create swap. Let's see if that helps.

**Comment by mlissner** [comment]
OK, the guide here is pretty solid (as usual): https://www.digitalocean.com/community/tutorials/how-to-add-swap-space-on-ubuntu-20-04

I created a 20GB swap. According to the [docker docs](https://docs.docker.com/config/containers/resource_constraints/#--memory-swap-details) the solr container should be able to use this automatically since the `--memory-swap` setting is not configured:

> If --memory-swap is unset, and --memory is set, the container can use as much swap as the --memory setting, if the host container has swap memory configured. For instance, if --memory="300m" and --memory-swap is not set, the container can use 600m in total of memory and swap.

**Comment by mlissner** [comment]
More downtime last night, so I gave it another 50GB swapfile. I could remove the smaller one, but it's not causing any trouble. Let's see how *this* goes.

**Comment by mlissner** [comment]
OK, I've been doing a lot of research and work on this the past couple days. A few notes:

1. We can't connect our current EFS volume to multiple instances. We are using gp3, and you have to use io1 or io2 to do that. We could convert to io2 (I think), and then turn on multi-attach (I think), but to turn on multi-attach, you have to disconnect the volume, so at that point, we might as well just connect it to the new machine rather than connect it to multiple machines.

2. Our solr docker image, thankfully, is already built for arm, so that's good. 

3. We should probably move to opensearch eventually. It'll make this kind of day less common, but it'll bring a bunch of cost and complexity too. Before we do this, we should probably just set up Elastic Search on our own server and see how we like it.

Here's my understanding for how to do this:

1. Set up a new instance. 
   - Create it from scratch mirroring existing instance
   - [Install/test docker](https://docs.docker.com/engine/install/ubuntu/)
   - Install courtlistener-solr per README.md
   - docker pull freelawproject/solr:latest

1. Configure the deployment in k8s to use the new IP address and ensure connection works.
   - Installing ping and then using that is a simple way to check. The firewall is pretty tightened down against other protocols, but it's open to ping.

3. Stop Solr on old instance
4. Dismount the volume from the old instance
5. [Attach](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-attaching-volume.html) and [mount](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-using-volumes.html) the volume on the new instance (this worked, but the old solr instance crashed when the disk was removed from it and couldn't be restored.)
6. Start the new solr
7. Re-deploy all of k8s to get the new settings.

After launch:

7. Make the volume [auto mount on reboot](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-using-volumes.html)

9. Get cloud monitoring working on new instance:
    - Official docs [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/install-CloudWatch-Agent-commandline-fleet.html), but [notes here helped](https://github.com/freelawproject/courtlistener/issues/2128#issuecomment-1220079978).
    - You have to give the EC2 instance the CloudWatchAgentServerRole IAM role
    - I set up a couple alerts as well. One for memory and one for storage.

11. Get log rotation working on new instance
 - I didn't do this, but I think it'll work by default and I couldn't find any notes about doing this in the past. I think it'll work by default because it's logging to the syslog, which should be rotated automatically. I have an alert for this, so I'm not too worried.

13. Terminate the old instance if everything is working. I did this, but I kept its root volume (`vol-0db23e75a89f44ec2`). We can delete it in a month or so if it's not needed anymore. If we want to make a new instance to do Whisper or something, the size to do is m6a.4xlarge.

1. The old solr uses a static IP, but I don't know why. I don't think it's needed. I didn't set the new server up with this and deleted the old IP. I think when we were setting up the cluster originally, we used this IP and instance to move data around.

3. Once launched, we can re-enable troller to send to solr. Done.

**Comment by mlissner** [comment]
Well, Solr ran out of memory again today. That's a bummer. The sooner we can move to Elastic, the better. A couple observations though:

1. The memory jumped a lot all at once. I think this was because somebody made a lot of queries, which moved a ton of data into memory:

    ![image](https://github.com/freelawproject/courtlistener/assets/236970/b776b5c0-f61c-491a-b296-643c37f37e36)

2. I created 124GB of swap space per the DO guide.

3. Our memory alert didn't go off because the disk cache was using a bunch of memory. We were looking for usage to be above 90%, but it never got there because the disk cache was using about 15%. I lowered the threshold for the alert to 80%.

I think these changes should buy us some time, but we need a new architecture that doesn't gobble up so much RAM.
